pkg/cdi/spec.go

```diff
@@ -0,0 +1,170 @@
+package cdi
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+
+	oci "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/yaml"
+
+	specs "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+)
+
+var (
+	// ValidSpecVersions contain the valid CDI Spec versions.
+	ValidSpecVersions = map[string]struct{}{
+		"":      {},
+		"0.1.0": {},
+		"0.2.0": {},
+	}
+)
+
+// Spec represents a CDI Spec in a Cache.
+type Spec struct {
+	*specs.Spec
+	path     string
+	priority int
+	vendor   string
+	class    string
+	devices  map[string]*Device
+}
+
+// LoadSpec loads a Spec from the filesystem.
+func LoadSpec(path string, priority int) (*Spec, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "failed to read CDI Spec")
+	}
+
+	spec, err := parseSpecData(data)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewSpec(spec, path, priority)
+}
+
+// NewSpec creates a new Spec with the given CDI Spec data.
+func NewSpec(s *specs.Spec, path string, priority int) (*Spec, error) {
+	spec := &Spec{
+		Spec:     s,
+		path:     filepath.Clean(path),
+		priority: priority,
+		devices:  make(map[string]*Device),
+	}
+
+	spec.vendor, spec.class = ParseQualifier(spec.Spec.Kind)
+
+	if err := spec.validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid CDI Spec")
+	}
+
+	for _, d := range spec.Devices {
+		device, err := newDevice(spec, d)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to add device %q", d.Name)
+		}
+		if _, ok := spec.devices[d.Name]; ok {
+			return nil, errors.Errorf("multiple devices %q defined", d.Name)
+		}
+		spec.devices[d.Name] = device
+	}
+
+	return spec, nil
+}
+
+// parseSpecData parses the given data into a CDI Spec.
+func parseSpecData(data []byte) (*specs.Spec, error) {
+	spec := &specs.Spec{}
+	err := yaml.UnmarshalStrict(data, spec)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse CDI Spec")
+	}
+
+	return spec, nil
+}
+
+// GetPath returns the path the Spec was loaded from.
+func (s *Spec) GetPath() string {
+	return s.path
+}
+
+// GetVendor returns the vendor for the Spec.
+func (s *Spec) GetVendor() string {
+	return s.vendor
+}
+
+// GetClass returns the device class for the Spec.
+func (s *Spec) GetClass() string {
+	return s.class
+}
+
+// ListDevices list devices in the Spec.
+func (s *Spec) ListDevices(f *Filter) ([]string, error) {
```

Comment:
That was actually unintentional on my side. Well spotted. I removed filtering from the basic function and added a *Matching* function instead which does it correctly. But since we agreed about getting rid of filtering/matching for now, I need to drop those bits altogether and update again.